src/vcpkg/base/downloads.cpp

```diff
@@ -83,12 +121,78 @@ namespace vcpkg
         }
     }
 
+    // On redirect, WinHttp API apparently requires to set again proxy credentials.
+    // Having this callback looks like the only way to achieve this.
+    void set_proxy_credentials_on_redirect(HINTERNET hInternet,
+                                           DWORD_PTR dwContext,
+                                           DWORD dwInternetStatus,
+                                           LPVOID /*lpvStatusInformation*/,
+                                           DWORD /*dwStatusInformationLength*/)
+    {
+        if (dwInternetStatus == WINHTTP_CALLBACK_STATUS_REDIRECT)
+        {
+            ProxyCredentials** ctxt_ptr_ptr = (ProxyCredentials**)dwContext;
+            ProxyCredentials& ctxt = **ctxt_ptr_ptr;
+            WinHttpSetCredentials(hInternet,
+                                  WINHTTP_AUTH_TARGET_PROXY,
+                                  WINHTTP_AUTH_SCHEME_BASIC,
+                                  ctxt.username.data(),
+                                  ctxt.password.data(),
+                                  NULL);
+        }
+    }
+
+    // Only purpose of this class is to manage the lifetime of a request context
+    // which needs to be provided as a "DWORD_PTR" to WinHttpSendRequest
+    // (pointer to a pointer-sized variable).
```

Comment:
`DWORD_PTR` is a pointer-sized variable itself. It models `uintptr_t` not `uintptr_t*`. You should be able to pass T* directly.